src/utils.jl

```diff
@@ -28,20 +28,50 @@ else
     const _getproperty = getproperty
 end
 
-function _foreachfield(names, xs)
+function _sstuple(::Type{<:NTuple{N, Any}}) where {N}
+    ntuple(j->Symbol(j), N)
+end
+
+function _sstuple(::Type{NT}) where {NT<:NamedTuple}
+    _map_params(x->_sstuple(staticschema(x)), NT)
+end
+
+function _getcolproperties!(exprs, s, es=[])
+    if typeof(s) <: Symbol
+        push!(exprs, es)
+        return
+    end
+    for key in keys(s)
+        _getcolproperties!(exprs, getproperty(s,key), vcat(es, key))
+    end
+end
+
+@generated function foreachfield(::Type{T}, f, xs...) where {T<:Tup}
```

Comment:
This introduces a lot of code complexity, while the original implementation is just a few lines of code: I'm not completely sure I understand conceptually what this refactor is trying to do. Is there an intuitive explanation as to way the original `foreachfield` is not suitable for CUDAnative while this implementation works?